### PR TITLE
key, release/resource/status: implement status resource

### PR DESCRIPTION
**Pull request by kopiczko**
Towards https://github.com/giantswarm/giantswarm/issues/5278.

This fills the Status.Cycle.Phase which will be used by the app resource
to make decisions about release components.

### CODE REVIEW

File: service/controller/release/resource/status/create.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	var releaseCycleCR *releasev1alpha1.ReleaseCycle
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding ReleaseCycle CR %#q in namespace %#q", cr.Name, cr.Namespace))
+
+		// Release CR and corresponding ReleaseCycle CR have the same name and namespace.
+		releaseCycleCR, err = r.g8sClient.ReleaseV1alpha1().ReleaseCycles(cr.Namespace).Get(cr.Name, metav1.GetOptions{})
```
Comment (xh3b4sd) [inline service/controller/release/resource/status/create.go]:
```suggestion
		releaseCycleCR, err = r.g8sClient.ReleaseV1alpha1().ReleaseCycles(cr.GetNamespace()).Get(cr.GetName(), metav1.GetOptions{})
```

File: service/controller/release/resource/status/create.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+		} else {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found ReleaseCycle CR %#q in namespace %#q", cr.Name, cr.Namespace))
+		}
+	}
+
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding if custom resource status needs to be updated")
+
+		if releaseCycleCR == nil && cr.Status.Cycle.Phase == releasev1alpha1.CyclePhaseUpcoming {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "found custom resource status does not need to be updated")
```
Comment (xh3b4sd) [inline service/controller/release/resource/status/create.go]: There is a reason log missing.

File: service/controller/release/resource/status/create.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+			return nil
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "found custom resource status needs to be updated")
+	}
+
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding the latest version of custom resource")
+
+		cr, err = r.g8sClient.ReleaseV1alpha1().Releases(cr.Namespace).Get(cr.Name, metav1.GetOptions{})
```
Comment (xh3b4sd) [inline service/controller/release/resource/status/create.go]:
```suggestion
		cr, err = r.g8sClient.ReleaseV1alpha1().Releases(cr.GetNamespace()).Get(cr.GetName(), metav1.GetOptions{})
```

File: service/controller/release/resource/status/create.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "updating custom resource status")
+
+		if releaseCycleCR == nil {
+			cr.Status.Cycle.Phase = releasev1alpha1.CyclePhaseUpcoming
+		} else {
+			cr.Status.Cycle = releaseCycleCR.Spec
+		}
+
+		_, err = r.g8sClient.ReleaseV1alpha1().Releases(cr.Namespace).UpdateStatus(cr)
```
Comment (xh3b4sd) [inline service/controller/release/resource/status/create.go]:
```suggestion
		_, err = r.g8sClient.ReleaseV1alpha1().Releases(cr.GetNamespace()).UpdateStatus(cr)
```

File: service/controller/release/resource/status/create.go (hunk 1 above)
Comment (kopiczko) [reply service/controller/release/resource/status/create.go]: Are you big on that / have a reason to use getter methods here? I personally prefer using fields directly. Less typing less parenthesis looks a bit clearer.

Comment (xh3b4sd) [reply service/controller/release/resource/status/create.go]: Not big on it. But they are there for a reason.

Comment (kopiczko) [reply service/controller/release/resource/status/create.go]: I think there reason for them to be is fulfilling interfaces which makes sense. I don't feel like they are intended to be used everywhere. I can be wrong though.

File: service/controller/release/resource/status/create.go (hunk 2 above)
Comment (kopiczko) [reply service/controller/release/resource/status/create.go]: Added below.

**Comment by kopiczko** [comment]
Added the logs and applied suggestions. Should at me if you'd like to improve them.

**Pull request merged**